### PR TITLE
Add high and low color settings for colormaps (HiLo mapping)

### PR DESCRIPTION
**Pull request by brisvag**
As suggested by @aganders3 in https://github.com/vispy/vispy/pull/2659#issuecomment-2786508704, based on the discussion in https://github.com/napari/napari/issues/6878.

Note that this includes changes from #2662. What do you think @aganders3?

You can try it out with this snippet:

```py
import sys
from vispy import scene, app
from vispy.color import get_colormap
import numpy as np


canvas = scene.SceneCanvas(keys='interactive', bgcolor='white')
canvas.size = 600, 800
view = canvas.central_widget.add_view()

img_data = np.zeros((500, 500), np.float32)
for i in range(10):
    img_data[i*50:(i+1)*50, :] = i

img_data[:, 20:60] = np.nan

image = scene.visuals.Image(img_data, 
                            cmap='fire',
                            clim=[1, 8],
                            interpolation='cubic', 
                            texture_format='auto')
view.add(image)

view.camera = scene.cameras.TurntableCamera()
view.camera.set_range()
canvas.show()
```

and then something like:

```
image.low_color = 'blue'
image.high_color = 'green'
```

### CODE REVIEW

File: vispy/visuals/image.py [hunk 1]
```diff
@@ -654,23 +673,18 @@ def _compute_bounds(self, axis, view):
...
-            # bad_color handling is part of the colormap already, not needed here,
-            # but we put a noop functions to simplify indexing in the different cases
-            fbad = Function('float noop (float data) {return data;}')
             fclim = Function(self._func_templates['clim_float'])
             fgamma = Function(self._func_templates['gamma_float'])
             # NOTE: red_to_luminance only uses the red component, fancy internalformats
             #   may need to use the other components or a different function chain
             fun = FunctionChain(
-                None, [Function(self._func_templates['red_to_luminance']), fbad, fclim, fgamma, Function(self.cmap.glsl_map)]
+                None, [Function(self._func_templates['red_to_luminance']), fclim, fgamma, Function(self.cmap.glsl_map)]
```
Comment (aganders3) [inline vispy/visuals/image.py:681]: We need to be careful changing this function chain, because there are places (at least in this file) it is referenced directly by index. This breaks the clim setter, for example.

File: vispy/color/colormap.py [hunk 2]
```diff
@@ -255,22 +259,75 @@ def __init__(self, colors=None):
...
+    def set_high_color(self, color=None, alpha=None):
+        if color is not None:
+            color = Color(color, alpha)
+            self._high_color = color
+            r, g, b, a = color.rgba
+
+            high_color_glsl = f"""// high_color_start
+            if (t == 1.0) {{
+                return vec4({r:.3f}, {g:.3f}, {b:.3f}, {a:.3f});
+            }} // high_color_end"""
```
Comment (aganders3) [inline vispy/color/colormap.py:298]: I had an issue with this when I hacked the demo in napari. It seems after normalizing some values at/above the upper clim can end up be below 1.0. I believe it's a floating point precision error, so I worked around it by making this comparison `if (t >= 1.0 - 1e-6)`. This is stable, but I'm not sure if there's an edge case it could be missing.

I haven't seen the same thing with the low color, but we should test to make sure.

I'll try to reproduce in just vispy.

Comment (aganders3) [reply vispy/color/colormap.py:298]: Try this, and note the flashing as you move the mouse up/down across certain values:
```python
from vispy import app, scene

import numpy as np


canvas = scene.SceneCanvas(keys="interactive")
canvas.size = 512, 512
canvas.show()
view = canvas.central_widget.add_view()
view.camera.rect = (0, 0, 512, 512)

image_data = np.linspace(0, 65535, 512 * 512, dtype=np.uint16).reshape((512, 512))
image_data[128:256, 128:256] = 65535
image_data[160:224, 160:224] = 65534

image = scene.visuals.Image(image_data, parent=view.scene)

image.cmap.set_high_color("red")
image.clim = (0.0, 34371.503496503494)


@canvas.events.mouse_move.connect
def on_mouse_move(event):
    image.clim = (
        0.0,
        float(event.pos[1] / canvas.size[1] * 65535),
    )
    print(image.clim)


if __name__ == "__main__":
    app.run()
```

**Comment by aganders3** [comment]
I left some comments on some minor things, but in general I like this API!

My only other nitpick is regex feels like overkill here, but I leave that up to you :).

File: vispy/visuals/image.py (hunk 1 above)
Comment (brisvag) [reply vispy/visuals/image.py:681]: ah yeah, I forgot to change it back. This PR actually goes back to how it was in main before I messed up #2659 (getting fixed in #2662)

File: vispy/color/colormap.py (hunk 2 above)
Comment (brisvag) [reply vispy/color/colormap.py]: Awesome snippet :D Ok, I went with an even smaller epsilon and it seems to be doing just fine.

**Comment by brisvag** [comment]
Thanks for reviewing! Fixed the issues.

I find regex cleaner here than just string methods actually, especially to make sure we handle all possible combination of adding/removing/readding/reremoving of bad, high, and low color.

**Comment by brisvag** [comment]
Updated and added some tests! What do you think @aganders3? @melonora you probably have opinions too!

**Comment by aganders3** [comment]
Tests pass locally (macOS) for me as well. I can try on my linux machine later today.

**Comment by aganders3** [comment]
Ahh, I think this is a test pollution problem. Colormaps get reused by vispy - not quite singletons necessarily but the changes can persist.

I think that raises questions about this API in general... I would find it surprising to set `image.cmap = "grays"` and get a hi-lo map because some _other_ visual had set grays with indicators.

**Comment by brisvag** [comment]
Aaah, right!

Mmm... I guess it could be argued either way. However, I don't see much benefit to having colormaps being singletons, other than comparison with `if colormap is Fire`. This sounds more like a "it's like that because there was no problem at the time" rather than "it should be like that". 

@djhoese what do you think? Is this just a bad api and we should rethink it? Or could we just change `get_colormap` to return new objects instead of the same one every time? Note that the work I just finished on nan handling in #2659 suffers from the same problem.

@jni I feel like you might have opinions on this as well :P

**Comment by melonora** [comment]
Tests pass locally for me on windows as well. I am excited about this PR and getting HiLo into napari (follow up PR to create FIJI like HiLo in vispy). 

That being said I think I do agree with @aganders3 here. Talking with @brisvag, even though downstream in napari this might not be a problem, within vispy itself it could be. From experience with other viewers people see HiLo really as a different colormap

**Comment by brisvag** [comment]
So, to summarize you suggest to:

- remove the setters on the `ImageVisual` (and if someone tries to set high and low colors on an existing image colormap via `Image.cmap.set_low_color()` and nothing happens, too bad)
- Create a `HiLo` colormap which is just `Grays` with low blue and high red
- Leave the rest as is (so if someone wants to create other colormaps with different high and low colors, they can do it manually)

Correct?

**Comment by brisvag** [comment]
Done! Well, making the colormap was easy so I made it here as well :P

**Comment by melonora** [comment]
LGTM, can't approve as I am not a maintainer but I am ok with merging this.

**Review by aganders3 — APPROVED** [review]
LGTM as well!

I'm a little unclear on what you mean by this:
> if someone tries to set high and low colors on an existing image colormap via `Image.cmap.set_low_color()` and nothing happens, too bad

I think this will work, right? The confusing part will be if they have two images using the same colormap, that will (likely) set the low color on _both_.

**Comment by djhoese** [comment]
It looks like there was a lot more discussion after @brisvag asked what my opinion was, so as long as I'm not the one who has to deal with issues resulting from this PR then :smile: :+1: 

That said, in my opinion, colormaps should be immutable. It is common in my experience to define colormaps with some text based format (CSV, YAML, even XML, etc). This static definition suggests that the in-memory version of the colormap should also be done that way. Also, typically the data itself has a "fill value" defined for it and that is either masked out before display/plotting/rendering or it is accounted for in the colormap itself. For example, having the lowest value of the colormap be black or transparent and then quickly switching to another color in the colormap so only fill values (ex. 0 of an uint8 array) are dark or transparent.

So some of my hesitation with this PR might be because I've never really used a colormap object with this functionality. That doesn't mean it is wrong or bad, just that I'm not really sure how I feel. Feel free to merge this, but it still feels like colormap objects should be immutable/static.

**Comment by jni** [comment]
> For example, having the lowest value of the colormap be black or transparent and then quickly switching to another color in the colormap so only fill values (ex. 0 of an uint8 array) are dark or transparent.

My understanding is that this is basically what this PR implements (?) — for both values above and below the contrast limits — but that there is subtlety due to quantisation of image values at various stages of the pipeline, when in fact we only want values *outside* of the contrast limits range in these colours.

> That said, in my opinion, colormaps should be immutable

indeed, but they are not — nothing is stopping me from modifying the color array of a colormap object, afaik. (?) But we shouldn't, which is why @brisvag added HiLo here — so no one needs to modify things to get it.

btw @brisvag

> and if someone tries to set high and low colors on an existing image colormap via Image.cmap.set_low_color() and nothing happens, too bad

I am with @aganders3 here, why wouldn't this work?

**Comment by brisvag** [comment]
> I'm a little unclear on what you mean by this:
> 
> > if someone tries to set high and low colors on an existing image colormap via `Image.cmap.set_low_color()` and nothing happens, too bad
> 
> I think this will work, right? The confusing part will be if they have two images using the same colormap, that will (likely) set the low color on _both_.

No, this won't work. The reason why I added the setters in `ImageVisual` in the first place was to deal with this specific issue. 

This is because the `color_transform` glsl code needs to be regenerated whenever something changes in the colormap, clim, or gamma settings. (By the way, I'm not sure why these are not handled using uniforms, but I've been down a similar rabbithole before and it's not a trivial task to switch to that system).

So in the current state, if you do `image.cmap.set_low_color(...)`, nothing happens. You'll have to somehow trigger a colortransform rebuild for it to take effect.

> it still feels like colormap objects should be immutable/static.

Yeah I think I'm convinced. Note however that colormaps already are technically mutable in vispy, just by changing the very public (and obvious) `cmap.colors` attribute. So I woulnd't go out of my way to try and actually make it hard to mutate more that it is. We can simply pre-generate colormaps (like it's been done so far) that also have high-low values, and leave the ability to create custom one if users want to.

**Comment by jni** [comment]
> You'll have to somehow trigger a colortransform rebuild for it to take effect.

Is this not something a colormap can trigger through an event? Or am I being too napari-brained? 😂 

> We can simply pre-generate colormaps (like it's been done so far) that also have high-low values, and leave the ability to create custom one if users want to.

👍 that's my take also.

**Comment by brisvag** [comment]
> Is this not something a colormap can trigger through an event? Or am I being too napari-brained? 😂

definitely napari-brained :P Just look at the code used by gamma and clim to see what I mean.

**Comment by brisvag** [comment]
Ok, so I went the static colormap route, and the code is a lot simpler because of it. Now bad/high/low colors are simply attributes of the class just like the colormap `colors` themselves, so everything follows the same API.

Note that while I included `bad_color` in these chages, differently from high/low we *always* map it, because we always want to handle NaNs in some way (by default they are transparent).

**Comment by brisvag** [comment]
Updated snippet:

```py
import sys
from vispy import scene, app
from vispy.color import get_colormap
import numpy as np


canvas = scene.SceneCanvas(keys='interactive', bgcolor='white')
canvas.size = 600, 800
view = canvas.central_widget.add_view()

img_data = np.zeros((500, 500), np.float32)
for i in range(10):
    img_data[i*50:(i+1)*50, :] = i

img_data[:, 20:60] = np.nan

image = scene.visuals.Image(img_data, 
                            cmap='HiLo',
                            clim=[0, 9],
                            interpolation='cubic', 
                            texture_format='auto')
view.add(image)

view.camera = scene.cameras.TurntableCamera()
view.camera.set_range()
canvas.show()
```

File: vispy/color/colormap.py [hunk 3]
```diff
@@ -403,7 +451,8 @@ class Colormap(BaseColormap):
 
     """
 
-    def __init__(self, colors, controls=None, interpolation='linear'):
+    def __init__(self, colors, controls=None, interpolation='linear',
```
Comment (aganders3) [inline vispy/color/colormap.py:454]: super-nit: consider making the new parameters kw-only for clarity and future proofing.

**Comment by aganders3** [comment]
I think this looks good! I agree it's cleaner, and the extra steps to make a custom HiLo map seem fine to me as it's a somewhat niche application.

One weird thing I thought of: the way you're adding the code to the `glsl_map` assumes it starts with something like:
```
vec4 grays(float t) {
   ...
```

I don't think this will necessarily always be the case, but I think existing code makes similar assumptions so maybe this is okay. I think it just means that some colormaps may not be suitable for converting to HiLo versions this way.

Otherwise LGTM!

**Comment by brisvag** [comment]
Yeah, but the pattern is very established (for the record, `grays` is not included , just `float t)`, and fixing it would require moving the the fancy shader compilation system which is a big lift. I think we're good like this :)
